transport/grpc/header.go

```diff
@@ -0,0 +1,85 @@
+package grpc
+
+import (
+	"strings"
+
+	"github.com/yarpc/yarpc-go/transport"
+	"google.golang.org/grpc/metadata"
+)
+
+// headerMapper converts gRCP Metadata to and from transport headers.
+type headerMapper struct{ Prefix string }
+
+var (
+	applicationHeaders = headerMapper{ApplicationHeaderPrefix}
+)
+
+// toGRPCMetadata converts transport headers into gRPC Metadata.
+//
+// Headers are read from 'from' and written to 'to'. The final header collection
+// is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func (hm headerMapper) ToGRPCMetadata(from transport.Headers, to metadata.MD) metadata.MD {
+	if to == nil {
+		to = make(metadata.MD, from.Len())
+	}
+	for k, v := range from.Items() {
+		Headers(to).Add(hm.Prefix+k, v)
+	}
+	return to
+}
+
+// fromGRPCMetadata converts GRPC Metadata to transport headers.
+//
+// Headers are read from 'from' and written to 'to'. The final header collection
+// is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func (hm headerMapper) FromGRPCMetadata(from metadata.MD, to transport.Headers) transport.Headers {
+	prefixLen := len(hm.Prefix)
+	for k := range from {
+		if strings.HasPrefix(k, hm.Prefix) {
+			key := k[prefixLen:]
+			to = to.With(key, Headers(from).Get(k))
+		}
+		// Note: undefined behavior for multiple occurrences of the same header
+	}
+	return to
+}
+
+// Headers is a convenience type for converting Header information safely
+type Headers map[string][]string
```

Comment:
Hmm, yeah this didn't need to be external, it was added as a simplification step for working on metadata headers, but we can just define the functions on metadata headers themselves.